runtime/src/rent_collector.rs

```diff
@@ -93,42 +93,20 @@ impl RentCollector {
         rent_for_sysvars: bool,
         filler_account_suffix: Option<&Pubkey>,
     ) -> u64 {
-        if !self.should_collect_rent(address, account, rent_for_sysvars)
-            || account.rent_epoch() > self.epoch
-            || crate::accounts_db::AccountsDb::is_filler_account_helper(
-                address,
-                filler_account_suffix,
-            )
+        if self.can_skip_rent_collection(address, account, rent_for_sysvars, filler_account_suffix)
         {
-            0
-        } else {
-            let (rent_due, exempt) = self.get_rent_due(account);
-
-            if exempt || rent_due != 0 {
-                if account.lamports() > rent_due {
-                    account.set_rent_epoch(
-                        self.epoch
-                            + if exempt {
-                                // Rent isn't collected for the next epoch
-                                // Make sure to check exempt status later in current epoch again
-                                0
-                            } else {
-                                // Rent is collected for next epoch
-                                1
-                            },
-                    );
-                    let _ = account.checked_sub_lamports(rent_due); // will not fail. We check above.
-                    rent_due
-                } else {
-                    let rent_charged = account.lamports();
-                    *account = AccountSharedData::default();
-                    rent_charged
-                }
-            } else {
-                // maybe collect rent later, leave account alone
-                0
-            }
+            return 0;
         }
+
+        let begin_lamports = account.lamports();
+        let rent_due = self.get_rent_due(account);
+
+        self.handle_rent_epoch(account, &rent_due);
+        self.handle_rent_amount(account, &rent_due);
+        self.handle_account_cleanup(account);
```

Comment:
I personally find this refactor harder to understand. It's not clear what each of these functions do based on their name. I like the `begin_lamports - end_lamports` change though. I think the rest can be inlined or be done in a single function

Reply:
Done!